File: golang/flatten_binary_tree.go
func flatten(root *TreeNode) {
	helper(root)
}

func helper(node *TreeNode) *TreeNode {
	// base case
	if node == nil {
		return nil
	}

	// find right first bc overwrite
	right := helper(node.Right)

	node.Right = helper(node.Left)
	node.Left = nil // remove left branch
	current := node
	for current.Right != nil {
		current = current.Right
	}
	current.Right = right

	return node
}

File: typescript/flatten_binary_tree.ts
function flatten(root: TreeNode | null): void {
	helper(root)
}

function helper(node: TreeNode): TreeNode {
	// base case
	if (node == null) return null

	// save right side, overwrite with left subtree
	let right = helper(node.right)
	node.right = helper(node.left)

	// cutoff left subtree
	node.left = null

	// traverse right side, attach to end
	let current = node
	while (current.right != null) {
		current = current.right
	}
	current.right = right

	return node
}
